### PR TITLE
add powershell cli script

### LINKED ISSUE #53764: Starting VSCode in the current directory fails if using PowerShell on a network share
Issue Type: <b>Bug</b>

When using Powershell on an unmapped network share, you cannot start VSCode using the command "code ." 

You'll get the error:
CMD.EXE was started with the above path as the current directory.
UNC paths are not supported.  Defaulting to Windows directory.

There are ways to work around, but it is less fast then being able to use "start powershell here" from explorer, and then typing "code .". 

VS Code version: Code 1.25.0 (0f080e5267e829de46638128001aeb7ca2d6d50e, 2018-07-05T13:11:58.697Z)
OS version: Windows_NT x64 10.0.16299

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i7-6600U CPU @ 2.60GHz (4 x 2808)|
|GPU Status|2d_canvas: enabled<br>flash_3d: enabled<br>flash_stage3d: enabled<br>flash_stage3d_baseline: enabled<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>native_gpu_memory_buffers: disabled_software<br>rasterization: disabled_software<br>video_decode: enabled<br>video_encode: enabled<br>vpx_decode: unavailable_software<br>webgl: enabled<br>webgl2: enabled|
|Memory (System)|7.93GB (1.22GB free)|
|Process Argv|C:\Program Files\Microsoft VS Code\Code.exe .|
|Screen Reader|no|
|VM|0%|

</details><details><summary>Extensions (9)</summary>

Extension|Author (truncated)|Version
---|---|---
pascal|ale|0.11.0
matlab|Gim|0.9.0
innosetup|idl|1.0.3
restructuredtext|lex|65.0.0
python|ms-|2018.6.0
cpptools|ms-|0.17.6
csharp|ms-|1.15.2
PowerShell|ms-|1.7.0
vscode-docker|Pet|0.0.27

</details>

### PR DESCRIPTION
**Pull request by jgbright**
A PowerShell script for launching code is a good way to deal with this.  It can deal with launching from an unmapped network share. There are still a couple of challenges to deal with.

1. The script needs to cope with quoted arguments.  For example, a user might want to run `code "c:\folder with spaces in the name"`.
2. The script should show cli output produced by code and it should wait until that output is complete before returning control to the caller.  For example, a user might want to run `code --list-extensions`.

One way to build the script is to use Start-Process with the ArgumentList parameter.  This does not work well with quoted arguments.  Another way is to use the call operator (&).  This works well with quoted arguments, but does not wait for code.exe to complete before returning control to the OS.  This results in weird behavior when running a command like `code --list-extensions` which writes to stdout and exits.  We can deal with this by piping output from the call operator to `Write-Output` or `Write-Host`.  I don't know if this will correctly deal with data written to stderr.

Fixes #53764

Here's a gif of what happens when the cli script for launch code does not wait for code to exit before returning.

code-bad-cli-handling

### CODE REVIEW

**Comment by jgbright** [comment]
I was thinking about how to test this change and ran through a few test cases.

# Tests
## Exit code
The exit code returned by code.exe is correctly returned by the script.

### cmd.exe on Windows 10
```
C:\repos\VSCode-win32-x64\bin>.\code-oss.cmd --file-write file-that-does-not-exist.txt
Using --file-write with invalid arguments.

C:\repos\VSCode-win32-x64\bin>echo %errorlevel%
1

C:\repos\VSCode-win32-x64\bin>
```

### PowerShell v6.1 on Windows 10
```
PS C:\repos\VSCode-win32-x64\bin> .\code-oss.ps1 --file-write file-that-does-not-exist.txt
Using --file-write with invalid arguments.
PS C:\repos\VSCode-win32-x64\bin> $LASTEXITCODE
1
PS C:\repos\VSCode-win32-x64\bin>
```


## Input stream

Piping text to stdin works with code.cmd, but not with code.ps1.  With code.cmd, the result is the app works as expected.  Code opens a new temp file with "Hello World" as the text.  With code.cmd, code opens, but no files are opened.

### cmd.exe on Windows 10
```
echo "Hello World" | code-oss.cmd -
```

### PowerShell v6.1 on Windows 10
```
PS C:\repos\VSCode-win32-x64\bin> Write-Output "Hello World" | .\code-oss.ps1 -
PS C:\repos\VSCode-win32-x64\bin>
```

## Output stream (stdout)

Output stream works as expected in code.cmd and code.ps1.  The script retains control of the terminal until code.exe returns.


### cmd.exe on Windows 10
```
C:\repos\VSCode-win32-x64\bin>code-oss.cmd --version
1.28.0
9b18cb18e3b4b3dc20e4937d325ad48d1554e649
x64

C:\repos\VSCode-win32-x64\bin>
```

### PowerShell v6.1 on Windows 10
```
PS C:\repos\VSCode-win32-x64\bin> .\code-oss.ps1 --version
1.28.0
9b18cb18e3b4b3dc20e4937d325ad48d1554e649
x64
PS C:\repos\VSCode-win32-x64\bin>
```

## Output stream (stderr)

I don't know of a way to trigger vscode to write to stderr.  I can trigger an error by passing an invalid filename on the cli to `--file-write` and that responds the same way with both code.cmd and code.ps1.

## Interactive cli

I don't know if code ever goes into a mode where it collects input directly from the terminal.  If it does, I suspect code.ps1 as written will not handle it correctly.

**Comment by jgbright** [comment]
@joaomoreno Here is the relevant documentation.  I suppose the installer could set an alias in the user’s PowerShell profile.  If that’s considered intrusive, it could be an install option.  Something like, check this box to install “code” as an alias.  I don’t know if I’ve ever seen an app install an alias before.

https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_command_precedence?view=powershell-6

**Comment by joaomoreno** [comment]
@jgbright Just tested this. Powershell by default picks up the `code.ps1` script over the `code.bat` one, which is great.

But by default, people can't run powershell scripts, so they'd need to update their execution policies:

image

We can't force people to do this, unfortunately. And I don't want to break people launching Code from Powershell. Do you have a suggestion for this?

**Comment by jgbright** [comment]
I see what you mean.  If we sign the script, it would be runnable under all PowerShell execution policies except `Restricted`.  However, the default execution policy is `Restricted`.  If we need to support that, I don’t think this approach is viable.

Maybe we could add a PowerShell alias that maps `code` to run the bat file.  @joaomoreno What do you think of that?  I could work on another pull request.

https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_execution_policies?view=powershell-5.1#powershell-execution-policies

**Comment by joaomoreno** [comment]
Yup, would be interested in that!

**Comment by jgbright** [comment]
@joaomoreno I have updated the branch to also install a PowerShell alias.

**Comment by joaomoreno** [comment]
I've created a build, ran the user setup and got this flashing error:

image

**Comment by jgbright** [comment]
I think the `-Force` argument added in my last commit will fix this.

**Comment by joaomoreno** [comment]
image

**Review by joaomoreno — CHANGES_REQUESTED** [review]
Requesting fix.

**Comment by jgbright** [comment]
@joaomoreno I was able to reproduce the error and code a fix.  👍

File: build/win32/code.iss [hunk 1]
```diff
@@ -90,6 +91,14 @@ Name: "{userappdata}\Microsoft\Internet Explorer\Quick Launch\{#NameLong}"; File
...
 Filename: "{app}\{#ExeBasename}.exe"; Description: "{cm:LaunchProgram,{#NameLong}}"; Tasks: runcode; Flags: nowait postinstall; Check: ShouldRunAfterUpdate
 Filename: "{app}\{#ExeBasename}.exe"; Description: "{cm:LaunchProgram,{#NameLong}}"; Flags: nowait postinstall; Check: WizardNotSilent
+#if "user" == InstallTarget
+#define PowerShellProfilePropName "CurrentUserAllHosts"
+#else
+#define PowerShellProfilePropName "AllUsersAllHosts"
+#endif
+#define ExeBasenameNoSpaces StringChange(ExeBasename, ' ', '')
+
+Filename: powershell.exe; Parameters: "if (!(Test-Path alias:code)) {{ New-Item -ItemType Directory -Force -Path (Split-Path $PROFILE.{#PowerShellProfilePropName}) | Out-Null; '','Set-Alias -Name code -Value ''{app}\bin\{#ExeBasenameNoSpaces}.ps1'' -Description ''{#NameLong}''' | Add-Content -Force -LiteralPath $profile.{#PowerShellProfilePropName} }"; Description: Updating PowerShell profile; Tasks: addpowershellalias; Flags: runascurrentuser
```
Comment (joaomoreno) [inline build/win32/code.iss]: Shouldn't `alias:code` be scoped? Remember we have insiders, stable, etc.

**Review by joaomoreno — CHANGES_REQUESTED** [review]
I still get the following when attempting to run `code-insiders`:

Screenshot from 12 08 2019 14:19:39

I thought your alias idea would work around this?

**Comment by joaomoreno** [comment]
Closing this due to inactivity, sorry @jgbright.

**Pull request closed without merging**